### PR TITLE
Add a relative prefix to the date divider in the chat

### LINKED ISSUE #559: Date divider improvements
Currently the dates show as absolute and not much info, just:
- January 9, 2018
- January 8, 2018
- etc.

More useful would be relative dates and weekdays too:
- Yesterday, Tuesday January 9
- 2 days ago, Monday January 8
- After 7 days ago: A week ago, Monday January 1

Year is only necessary to be displayed when you are so much in the past that it’s more than 11 months ago:
- A year ago, Friday January 20, 2017

### PR DESCRIPTION
**Pull request by nickvergessen**
@jancborchardt it's not exactly the steps you mentioned, but instead its automatically done by momentjs.
I think it's good enough, but since "a month ago" is repeating with each day in the past, I'd also be fine with only showing relative until it's no longer displayed as "days ago".
No one needs relative dates for "7 months ago".

> ug-test_-_gesprach_-_nextcloud_-_2018-01-15_16 10 07

Fix #559

### CODE REVIEW

File: js/views/chatview.js [hunk 1]
```diff
@@ -242,6 +240,29 @@
...
+
+				if (dayOfYear === yesterday) {
+					relativePrefix = t('spreed', 'Yesterday');
+				} else {
+					relativePrefix = date.fromNow();
+				}
+			}
+
+
+			return relativePrefix + ', ' + OC.Util.formatDate(timestamp, 'LL');
```
Comment (danxuliu) [inline js/views/chatview.js]: The whole text should be internationalized, as depending on the locale a different separator (or even order of the prefix and date) could be needed.

Also please add a comment like _'LL' formats a localized date including day of month, month name and year_ to know what it does ;-)

File: js/views/chatview.js [hunk 2]
```diff
@@ -242,6 +240,29 @@
 			}
 		},
 
+		_getDateSeparator: function(timestamp) {
+			var date = moment(timestamp, 'x'),
+				today = moment(),
+				dayOfYear = OC.Util.formatDate(date, 'YYYY-DDD'),
+				dateOfToday = OC.Util.formatDate(today, 'YYYY-DDD');
```
Comment (danxuliu) [inline js/views/chatview.js]: `dateOfToday` -> `currentDayOfYear`/`dayOfYearNow`/`dayOfYearToday`/whatever including `dayOfYear` ;-)

File: js/views/chatview.js [hunk 3]
```diff
@@ -242,6 +240,29 @@
 			}
 		},
 
+		_getDateSeparator: function(timestamp) {
+			var date = moment(timestamp, 'x'),
+				today = moment(),
```
Comment (danxuliu) [inline js/views/chatview.js]: Maybe it would be better to call it `now` instead of `today`? Not really sure which one would be clearer, though.

**Comment by jancborchardt** [comment]
Looks nice! :) I think the relative dates of even the multiple months ago are useful when you scroll back.

@nickvergessen so now that the date display on the right of each message duplicates the divider info, can we just show the timestamp of the specific message there on the right? Then that part also has a shorter and predictable width.

**Comment by nickvergessen** [comment]
> bildschirmfoto von 2018-01-16 11-18-40

We are using relative dates now for the first 24h, afterwards it switches to Time (or should we include the date too, because the date divider might not be visible?
Or can we pin the last date divider with some css magic?

**Comment by danxuliu** [comment]
I have rewritten a little the commit history to fix a lint issue (ignore the red mark next to the latest commit, I restarted the build and [it worked fine](https://drone.nextcloud.com/nextcloud/spreed/416) ;-) ), add the review changes to the original commit and split adding the relative prefix from showing the date in the first message.

However, there are still some issues. Although the relative date is not set initially due to the use of the `live-relative-timestamp` CSS class it will be set anyway after 30 seconds. Also, the relative date in the header is not automatically updated, so when chatting near midnight you will end with two _Today_ headers. Probably it will be needed to use a custom function to update the relative time akin to [the one used in the server](https://github.com/nextcloud/server/blob/de5467811a52abcc16d0b536136e09f1b614d79c/core/js/js.js#L1666-L1671).

@nickvergessen 
>  should we include the date too, because the date divider might not be visible?

Not sure about the date; it may be too much visually, and if you need to find the date you can hover on the time... or, in touch screens, scroll back until you find the date header. That could be quite annoying, though, so I do not know.

In any case I would add back the seconds, at the very least to the tooltip and maybe to the text itself too, as otherwise it would not be possible to know them in touch screens.

>  Or can we pin the last date divider with some css magic?

I am not aware of any CSS-only way to pin the last divider and, although it could be a cool effect, for the time being I would not add it to keep the code simple (there are enough bugs to fix already ;-) ).

**Comment by nickvergessen** [comment]
Fixed the issue with `live-relative-timestamp`

> In any case I would add back the seconds, at the very least to the tooltip and maybe to the text itself too, as otherwise it would not be possible to know them in touch screens.

Done

> I am not aware of any CSS-only way to pin the last divider and, although it could be a cool effect, for the time being I would not add it to keep the code simple (there are enough bugs to fix already ;-) ).

Yeah, that's why I asked for some simple CSS, lets just keep it like that for now then.

**Comment by nickvergessen** [comment]
What do you mean?
after midnight it still says `1 day ago`, although it should be `2 days ago` now?

**Comment by nickvergessen** [comment]
Merging as discussed

**Pull request merged**